File: examples/registered/cpp/face_feature_demo/facedet_onnx.cpp
#include "facedet_onnx.h"

#include <onnxruntime_cxx_api.h>

#include <algorithm>
#include <array>
#include <cmath>
#include <cstdlib>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

#if defined(__linux__)
#include <limits.h>
#include <unistd.h>
#endif

namespace {

constexpr const char* kModelName = "face_detect.onnx";
constexpr float kConfidenceThreshold = 0.2f;
constexpr float kNMSThreshold = 0.45f;
constexpr int kTopK = 1000;
constexpr int kKeepTopK = 512;

struct DecodedFace {
  float score = 0.0f;
  float x1 = 0.0f;
  float y1 = 0.0f;
  float x2 = 0.0f;
  float y2 = 0.0f;
  std::array<float, 10> landmarks{};
};

float sigmoid(float x) {
  x = std::max(-88.3762626647949f, std::min(88.3762626647949f, x));
  return 1.0f / (1.0f + std::exp(-x));
}

float iou(const DecodedFace& a, const DecodedFace& b) {
  const float xx1 = std::max(a.x1, b.x1);
  const float yy1 = std::max(a.y1, b.y1);
  const float xx2 = std::min(a.x2, b.x2);
  const float yy2 = std::min(a.y2, b.y2);
  const float inter = std::max(0.0f, xx2 - xx1) * std::max(0.0f, yy2 - yy1);
  const float area_a = std::max(0.0f, a.x2 - a.x1) * std::max(0.0f, a.y2 - a.y1);
  const float area_b = std::max(0.0f, b.x2 - b.x1) * std::max(0.0f, b.y2 - b.y1);
  const float denom = area_a + area_b - inter;
  return denom > 0.0f ? inter / denom : 0.0f;
}

std::filesystem::path executableDir() {
#if defined(__linux__)
  std::array<char, PATH_MAX> buf{};
  const ssize_t n = readlink("/proc/self/exe", buf.data(), buf.size() - 1);
  if (n > 0) {
    buf[static_cast<size_t>(n)] = '\0';
    return std::filesystem::path(buf.data()).parent_path();
  }
#endif
  return std::filesystem::current_path();
}

std::filesystem::path findModelPath(const std::string& requested) {
  if (!requested.empty()) {
    std::filesystem::path p(requested);
    if (std::filesystem::exists(p)) return p;
  }
  if (const char* env = std::getenv("W_AGENT_FACEDET_ONNX")) {
    std::filesystem::path p(env);
    if (std::filesystem::exists(p)) return p;
  }
  const std::filesystem::path exe_dir = executableDir();
  const std::array<std::filesystem::path, 5> candidates = {
      exe_dir / kModelName,
      exe_dir.parent_path() / kModelName,
      exe_dir.parent_path().parent_path() / kModelName,
      std::filesystem::current_path() / kModelName,
      std::filesystem::path(__FILE__).parent_path() / kModelName,
  };
  for (const auto& p : candidates) {
    if (std::filesystem::exists(p)) return p;
  }
  throw std::runtime_error("face detector ONNX model not found: " + std::string(kModelName));
}

void preprocessBGR(const cv::Mat& bgr, std::vector<float>& out, int& rows, int& cols) {
  if (bgr.empty() || bgr.channels() != 3) {
    throw std::runtime_error("face detector input must be a non-empty BGR image");
  }
  const int height = bgr.rows;
  const int width = bgr.cols;
  rows = ((height - 1) / 32 + 1) * 16;
  cols = ((width - 1) / 32 + 1) * 16;
  out.assign(static_cast<size_t>(32) * rows * cols, 0.0f);
  for (int r = 0; r < rows; ++r) {
    for (int c = 0; c < cols; ++c) {
      for (int fy = -1; fy <= 1; ++fy) {
        const int src_y = r * 2 + fy;
        if (src_y < 0 || src_y >= height) continue;
        const unsigned char* row_ptr = bgr.ptr<unsigned char>(src_y);
        for (int fx = -1; fx <= 1; ++fx) {
          const int src_x = c * 2 + fx;
          if (src_x < 0 || src_x >= width) continue;
          const int offset = (fy + 1) * 3 + fx + 1;
          const unsigned char* p = row_ptr + src_x * 3;
          for (int ch = 0; ch < 3; ++ch) {
            out[static_cast<size_t>(offset * 3 + ch) * rows * cols + static_cast<size_t>(r) * cols + c] =
                static_cast<float>(p[ch]);
          }
        }
      }
    }
  }
}

void decodeLevel(
    const float* cls,
    const float* reg,
    const float* kps,
    const float* obj,
    const std::vector<int64_t>& shape,
    int stride,
    std::vector<DecodedFace>& out) {
  if (!cls || !reg || !kps || !obj || shape.size() != 4) return;
  const int h = static_cast<int>(shape[2]);
  const int w = static_cast<int>(shape[3]);
  const size_t plane = static_cast<size_t>(h) * w;
  for (int y = 0; y < h; ++y) {
    for (int x = 0; x < w; ++x) {
      const size_t idx = static_cast<size_t>(y) * w + x;
      const float score = std::sqrt(sigmoid(cls[idx]) * sigmoid(obj[idx]));
      if (score < kConfidenceThreshold) continue;
      const float prior_x = static_cast<float>(x * stride);
      const float prior_y = static_cast<float>(y * stride);
      const float cx = reg[idx] * stride + prior_x;
      const float cy = reg[plane + idx] * stride + prior_y;
      const float bw = std::exp(reg[2 * plane + idx]) * stride;
      const float bh = std::exp(reg[3 * plane + idx]) * stride;
      DecodedFace face;
      face.score = score;
      face.x1 = cx - bw * 0.5f;
      face.y1 = cy - bh * 0.5f;
      face.x2 = cx + bw * 0.5f;
      face.y2 = cy + bh * 0.5f;
      for (int point = 0; point < 5; ++point) {
        face.landmarks[point * 2] = kps[static_cast<size_t>(point * 2) * plane + idx] * stride + prior_x;
        face.landmarks[point * 2 + 1] = kps[static_cast<size_t>(point * 2 + 1) * plane + idx] * stride + prior_y;
      }
      out.push_back(face);
    }
  }
}

std::vector<DecodedFace> runNMS(std::vector<DecodedFace> faces) {
  std::stable_sort(faces.begin(), faces.end(), [](const DecodedFace& a, const DecodedFace& b) {
    return a.score > b.score;
  });
  if (kTopK > -1 && static_cast<int>(faces.size()) > kTopK) faces.resize(kTopK);

  std::vector<DecodedFace> kept;
  for (const auto& face : faces) {
    bool keep = true;
    for (const auto& prev : kept) {
      if (iou(face, prev) > kNMSThreshold) {
        keep = false;
        break;
      }
    }
    if (keep) kept.push_back(face);
    if (kKeepTopK > -1 && static_cast<int>(kept.size()) >= kKeepTopK) break;
  }
  return kept;
}

}  // namespace

struct FaceDetectorONNX::Impl {
  explicit Impl(const std::string& requested_model_path)
      : model_path(findModelPath(requested_model_path)),
        env(ORT_LOGGING_LEVEL_WARNING, "w-agent-facedet") {
    Ort::SessionOptions options;
    options.SetIntraOpNumThreads(1);
    options.SetInterOpNumThreads(1);
    options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
    session = std::make_unique<Ort::Session>(env, model_path.string().c_str(), options);

    Ort::AllocatorWithDefaultOptions allocator;
    input_name = session->GetInputNameAllocated(0, allocator).get();
    const size_t output_count = session->GetOutputCount();
    output_names.reserve(output_count);
    output_name_ptrs.reserve(output_count);
    for (size_t i = 0; i < output_count; ++i) {
      output_names.emplace_back(session->GetOutputNameAllocated(i, allocator).get());
    }
    for (const auto& name : output_names) output_name_ptrs.push_back(name.c_str());
    std::cerr << "[facedet] onnxruntime cpu model=" << model_path << std::endl;
  }

  std::filesystem::path model_path;
  Ort::Env env;
  std::unique_ptr<Ort::Session> session;
  std::string input_name;
  std::vector<std::string> output_names;
  std::vector<const char*> output_name_ptrs;
};

FaceDetectorONNX::FaceDetectorONNX(const std::string& model_path)
    : impl_(std::make_unique<Impl>(model_path)) {}

FaceDetectorONNX::~FaceDetectorONNX() = default;

std::vector<FaceCandidate> FaceDetectorONNX::detect(const cv::Mat& bgr) const {
  if (!impl_ || !impl_->session) return {};
  std::vector<float> tensor;
  int rows = 0;
  int cols = 0;
  preprocessBGR(bgr, tensor, rows, cols);

  std::array<int64_t, 4> input_shape = {1, 32, rows, cols};
  Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
  Ort::Value input = Ort::Value::CreateTensor<float>(
      memory_info,
      tensor.data(),
      tensor.size(),
      input_shape.data(),
      input_shape.size());
  const char* input_names[] = {impl_->input_name.c_str()};
  std::vector<Ort::Value> outputs = impl_->session->Run(
      Ort::RunOptions{nullptr},
      input_names,
      &input,
      1,
      impl_->output_name_ptrs.data(),
      impl_->output_name_ptrs.size());

  std::vector<DecodedFace> decoded;
  for (size_t level = 0; level < 3; ++level) {
    const size_t base = level * 4;
    if (outputs.size() <= base + 3) break;
    const auto shape = outputs[base].GetTensorTypeAndShapeInfo().GetShape();
    decodeLevel(
        outputs[base].GetTensorData<float>(),
        outputs[base + 1].GetTensorData<float>(),
        outputs[base + 2].GetTensorData<float>(),
        outputs[base + 3].GetTensorData<float>(),
        shape,
        level == 0 ? 8 : (level == 1 ? 16 : 32),
        decoded);
  }

  decoded = runNMS(std::move(decoded));
  std::vector<FaceCandidate> faces;
  faces.reserve(decoded.size());
  for (const auto& item : decoded) {
    FaceCandidate face;
    face.score = item.score;
    face.box = cv::Rect(
        static_cast<int>(item.x1),
        static_cast<int>(item.y1),
        static_cast<int>(item.x2 - item.x1),
        static_cast<int>(item.y2 - item.y1));
    for (int point = 0; point < 5; ++point) {
      face.landmarks.emplace_back(item.landmarks[point * 2], item.landmarks[point * 2 + 1]);
    }
    faces.push_back(std::move(face));
  }
  return faces;
}

FaceCandidate detectBestFaceONNX(const cv::Mat& bgr, FaceDetectorONNX& detector) {
  std::vector<FaceCandidate> faces = detector.detect(bgr);
  if (faces.empty()) {
    throw std::runtime_error("no face detected");
  }
  return *std::max_element(faces.begin(), faces.end(), [](const FaceCandidate& a, const FaceCandidate& b) {
    return a.score * std::max(1, a.box.area()) < b.score * std::max(1, b.box.area());
  });
}

cv::Mat alignFace(const cv::Mat& bgr, const FaceCandidate& face) {
  if (face.landmarks.size() < 2) {
    throw std::runtime_error("face landmarks missing");
  }
  cv::Point2f leftEye = face.landmarks[0];
  cv::Point2f rightEye = face.landmarks[1];
  cv::Point2f eyeCenter((leftEye.x + rightEye.x) * 0.5f, (leftEye.y + rightEye.y) * 0.5f);
  double angle = std::atan2(rightEye.y - leftEye.y, rightEye.x - leftEye.x) * 180.0 / CV_PI;
  double eyeDistance = cv::norm(rightEye - leftEye);
  double targetEyeDistance = 48.0;
  double scale = targetEyeDistance / std::max(eyeDistance, 1.0);
  cv::Mat rot = cv::getRotationMatrix2D(eyeCenter, angle, scale);
  rot.at<double>(0, 2) += 56.0 - eyeCenter.x;
  rot.at<double>(1, 2) += 44.0 - eyeCenter.y;
  cv::Mat aligned;
  cv::warpAffine(bgr, aligned, rot, cv::Size(112, 112), cv::INTER_LINEAR, cv::BORDER_REPLICATE);
  return aligned;
}

File: examples/registered/python/face_feature_demo/facedet_onnx.py
from __future__ import annotations

import math
from dataclasses import dataclass
from pathlib import Path

import cv2
import numpy as np
import onnxruntime as ort


MODEL_NAME = "face_detect.onnx"


@dataclass
class FaceCandidate:
    score: float
    box: tuple[int, int, int, int]
    landmarks: list[tuple[float, float]]


def sigmoid(values: np.ndarray) -> np.ndarray:
    return 1.0 / (1.0 + np.exp(-np.clip(values, -88.3762626647949, 88.3762626647949)))


def preprocess_bgr_to_libfacedet_tensor(bgr: np.ndarray) -> np.ndarray:
    height, width = bgr.shape[:2]
    rows = ((height - 1) // 32 + 1) * 16
    cols = ((width - 1) // 32 + 1) * 16
    out = np.zeros((1, 32, rows, cols), dtype=np.float32)
    for r in range(rows):
        for c in range(cols):
            for fy in (-1, 0, 1):
                src_y = r * 2 + fy
                if src_y < 0 or src_y >= height:
                    continue
                for fx in (-1, 0, 1):
                    src_x = c * 2 + fx
                    if src_x < 0 or src_x >= width:
                        continue
                    offset = (fy + 1) * 3 + fx + 1
                    out[0, offset * 3 : offset * 3 + 3, r, c] = bgr[src_y, src_x, :3]
    return out


def iou(a: tuple[float, float, float, float], b: tuple[float, float, float, float]) -> float:
    ax1, ay1, ax2, ay2 = a
    bx1, by1, bx2, by2 = b
    xx1 = max(ax1, bx1)
    yy1 = max(ay1, by1)
    xx2 = min(ax2, bx2)
    yy2 = min(ay2, by2)
    inter = max(0.0, xx2 - xx1) * max(0.0, yy2 - yy1)
    area_a = max(0.0, ax2 - ax1) * max(0.0, ay2 - ay1)
    area_b = max(0.0, bx2 - bx1) * max(0.0, by2 - by1)
    denom = area_a + area_b - inter
    return inter / denom if denom > 0 else 0.0


class FaceDetectorONNX:
    def __init__(self, model_path: Path | None = None, confidence_threshold: float = 0.2, nms_threshold: float = 0.45, topk: int = 1000, keep_topk: int = 512):
        self.model_path = Path(model_path or Path(__file__).with_name(MODEL_NAME))
        if not self.model_path.exists():
            raise FileNotFoundError(f"face detector ONNX model not found: {self.model_path}")
        self.confidence_threshold = confidence_threshold
        self.nms_threshold = nms_threshold
        self.topk = topk
        self.keep_topk = keep_topk
        self.session = ort.InferenceSession(str(self.model_path), providers=["CPUExecutionProvider"])
        self.input_name = self.session.get_inputs()[0].name

    def detect(self, bgr: np.ndarray) -> list[FaceCandidate]:
        tensor = preprocess_bgr_to_libfacedet_tensor(bgr)
        outputs = self.session.run(None, {self.input_name: tensor})
        levels = [
            (outputs[0], outputs[1], outputs[2], outputs[3], 8),
            (outputs[4], outputs[5], outputs[6], outputs[7], 16),
            (outputs[8], outputs[9], outputs[10], outputs[11], 32),
        ]
        candidates: list[tuple[float, tuple[float, float, float, float], list[tuple[float, float]]]] = []
        for cls, reg, kps, obj, stride in levels:
            candidates.extend(self._decode_level(cls[0], reg[0], kps[0], obj[0], stride))
        candidates.sort(key=lambda item: item[0], reverse=True)
        if self.topk > -1:
            candidates = candidates[: self.topk]

        kept: list[tuple[float, tuple[float, float, float, float], list[tuple[float, float]]]] = []
        for candidate in candidates:
            if all(iou(candidate[1], prev[1]) <= self.nms_threshold for prev in kept):
                kept.append(candidate)
            if self.keep_topk > -1 and len(kept) >= self.keep_topk:
                break

        faces: list[FaceCandidate] = []
        for score, box, landmarks in kept:
            x1, y1, x2, y2 = box
            faces.append(
                FaceCandidate(
                    score=score,
                    box=(int(x1), int(y1), int(x2 - x1), int(y2 - y1)),
                    landmarks=[(float(x), float(y)) for x, y in landmarks],
                )
            )
        return faces

    def _decode_level(self, cls: np.ndarray, reg: np.ndarray, kps: np.ndarray, obj: np.ndarray, stride: int):
        scores = np.sqrt(sigmoid(cls[0]) * sigmoid(obj[0]))
        ys, xs = np.where(scores >= self.confidence_threshold)
        out = []
        for y, x in zip(ys, xs):
            prior_x = float(x * stride)
            prior_y = float(y * stride)
            cx = float(reg[0, y, x]) * stride + prior_x
            cy = float(reg[1, y, x]) * stride + prior_y
            width = math.exp(float(reg[2, y, x])) * stride
            height = math.exp(float(reg[3, y, x])) * stride
            box = (cx - width * 0.5, cy - height * 0.5, cx + width * 0.5, cy + height * 0.5)
            landmarks = []
            for idx in range(5):
                landmarks.append(
                    (
                        float(kps[idx * 2, y, x]) * stride + prior_x,
                        float(kps[idx * 2 + 1, y, x]) * stride + prior_y,
                    )
                )
            out.append((float(scores[y, x]), box, landmarks))
        return out


def align_face(bgr: np.ndarray, face: FaceCandidate) -> np.ndarray:
    if len(face.landmarks) < 2:
        raise ValueError("face landmarks missing")
    left_eye = np.asarray(face.landmarks[0], dtype=np.float32)
    right_eye = np.asarray(face.landmarks[1], dtype=np.float32)
    eye_center = (left_eye + right_eye) * 0.5
    angle = math.degrees(math.atan2(float(right_eye[1] - left_eye[1]), float(right_eye[0] - left_eye[0])))
    eye_distance = float(np.linalg.norm(right_eye - left_eye))
    scale = 48.0 / max(eye_distance, 1.0)
    rot = cv2.getRotationMatrix2D((float(eye_center[0]), float(eye_center[1])), angle, scale)
    rot[0, 2] += 56.0 - float(eye_center[0])
    rot[1, 2] += 44.0 - float(eye_center[1])
    return cv2.warpAffine(bgr, rot, (112, 112), flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_REPLICATE)


def detect_best_face(bgr: np.ndarray, detector: FaceDetectorONNX | None = None) -> FaceCandidate:
    faces = (detector or FaceDetectorONNX()).detect(bgr)
    if not faces:
        raise ValueError("no face detected")
    return max(faces, key=lambda face: face.score * max(1, face.box[2] * face.box[3]))
